ローカルファイル読込時にNullPointerExceptionが発生することがある

## 不具合の概要(Description of bug)

- SDKのバージョン(SDK version)
 - v2.2.3
- 不具合が発生した環境(Environment requirement)
 - Android 4.0 ~
- 不具合を再現させる手順・サンプルコード(Step for Reproduce bug / Sample code)
 - 端末の実行環境によってInstallation新規登録後の更新時にNullPointerExceptionが発生することがある
   - エラーが発生する実行環境には、ディスク容量の不足などが考えられます
    - 書込時に空のローカルファイルが作成され、読込時にNullPointerExceptionが発生している模様
 - [読込時の処理](https://github.com/NIFTYCloud-mbaas/ncmb_android/blob/v2.2.3/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBLocalFile.java#L52)にNullチェックを入れるか、[書込時の処理](https://github.com/NIFTYCloud-mbaas/ncmb_android/blob/v2.2.3/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBLocalFile.java#L33)に空ファイルが作られないようにするなどの修正が必要かと思われます
- エラーメッセージ(Error messages)
    ```
Caused by: java.lang.RuntimeException: java.lang.NullPointerException
   at com.nifty.cloud.mb.core.NCMBInstallation.getCurrentInstallation(NCMBInstallation.java:381)
   at mbaas.com.nifty.ncmbpushquickstart.MainActivity.onCreate(MainActivity.java:44)
   at android.app.Activity.performCreate(Activity.java:5231)
   at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1087)
   at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2159)
   at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2245)
   at android.app.ActivityThread.access$800(ActivityThread.java:135)
   at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1196)
   at android.os.Handler.dispatchMessage(Handler.java:102)
   at android.os.Looper.loop(Looper.java:136)
   at android.app.ActivityThread.main(ActivityThread.java:5017)
   at java.lang.reflect.Method.invokeNative(Native Method)
   at java.lang.reflect.Method.invoke(Method.java:515)
   at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:779)
   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:595)
   at dalvik.system.NativeStart.main(Native Method)

Caused by: java.lang.NullPointerException
   at org.json.JSONTokener.nextCleanInternal(JSONTokener.java:116)
   at org.json.JSONTokener.nextValue(JSONTokener.java:94)
   at org.json.JSONObject.<init>(JSONObject.java:155)
   at org.json.JSONObject.<init>(JSONObject.java:172)
   at com.nifty.cloud.mb.core.NCMBLocalFile.readFile(NCMBLocalFile.java:52)
   at com.nifty.cloud.mb.core.NCMBInstallation.getCurrentInstallation(NCMBInstallation.java:374)
```

--- a/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBInstallation.java
+++ b/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBInstallation.java
@@ -358,7 +358,8 @@ public class NCMBInstallation extends NCMBObject {
     /**
      * Get current installation object
      *
-     * @return Return NCMBInstallation object created from the contents read local file. if local file is empty, return new NCMBInstallation.
+     * @return NCMBInstallation object that is created from data that is saved to local file.<br>
+     * If local file is not available, it returns empty NCMBInstallation object
      */
     public static NCMBInstallation getCurrentInstallation() {
         try {

--- a/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBInstallation.java
+++ b/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBInstallation.java
@@ -358,7 +358,7 @@ public class NCMBInstallation extends NCMBObject {
     /**
      * Get current installation object
      *
-     * @return Installation
+     * @return Return NCMBInstallation object created from the contents read local file. if local file is empty, return new NCMBInstallation.
      */
     public static NCMBInstallation getCurrentInstallation() {
         try {

--- a/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBLocalFile.java
+++ b/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBLocalFile.java
@@ -34,6 +34,10 @@ class NCMBLocalFile {
             out.close();
         } catch (IOException e) {
             throw new RuntimeException(e);
+        } finally {
+            if(writeFile.exists() && writeFile.length() ==0){
+                writeFile.delete();
+            }
         }
     }
 

--- a/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBLocalFile.java
+++ b/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBLocalFile.java
@@ -45,7 +45,7 @@ class NCMBLocalFile {
      * Reading to a local file
      *
      * @param readFile read file instance
-     * @return file data
+     * @return file data. if file data is empty, return empty json
      */
     static JSONObject readFile(File readFile) {
         checkNCMBContext();

--- a/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBLocalFile.java
+++ b/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBLocalFile.java
@@ -34,10 +34,6 @@ class NCMBLocalFile {
             out.close();
         } catch (IOException e) {
             throw new RuntimeException(e);
-        } finally {
-            if(writeFile.exists() && writeFile.length() ==0){
-                writeFile.delete();
-            }
         }
     }
 

--- a/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBLocalFile.java
+++ b/ncmb-core/src/main/java/com/nifty/cloud/mb/core/NCMBLocalFile.java
@@ -15,18 +15,22 @@ import java.io.IOException;
  * LocalFile class
  */
 class NCMBLocalFile {
-    /** Base folder name */
+    /**
+     * Base folder name
+     */
     static final String FOLDER_NAME = "NCMB";
 
-    static File create(String fileName){
+    static File create(String fileName) {
         return new File(NCMB.getCurrentContext().context.getDir(FOLDER_NAME, Context.MODE_PRIVATE), fileName);
     }
+
     /**
      * Writing to a local file
+     *
      * @param writeFile write file instance
-     * @param fileData local file data
+     * @param fileData  local file data
      */
-    static void writeFile(File writeFile , JSONObject fileData){
+    static void writeFile(File writeFile, JSONObject fileData) {
         checkNCMBContext();
         try {
             FileOutputStream out = new FileOutputStream(writeFile);
@@ -39,15 +43,19 @@ class NCMBLocalFile {
 
     /**
      * Reading to a local file
+     *
      * @param readFile read file instance
      * @return file data
      */
-    static JSONObject readFile(File readFile){
+    static JSONObject readFile(File readFile) {
         checkNCMBContext();
         JSONObject json = new JSONObject();
         try {
             BufferedReader br = new BufferedReader(new FileReader(readFile));
             String information = br.readLine();
+            if (information == null) {
+                return json;
+            }
             br.close();
             json = new JSONObject(information);
         } catch (IOException | JSONException e) {
@@ -58,9 +66,10 @@ class NCMBLocalFile {
 
     /**
      * Delete the local file
+     *
      * @param deleteFile delete file instance
      */
-    static void deleteFile(File deleteFile){
+    static void deleteFile(File deleteFile) {
         checkNCMBContext();
         deleteFile.delete();
     }
@@ -68,7 +77,7 @@ class NCMBLocalFile {
     /**
      * null check in NCMBContext
      */
-    static void checkNCMBContext(){
+    static void checkNCMBContext() {
         if (NCMB.getCurrentContext() == null) {
             throw new RuntimeException("Please run the　NCMB.initialize.");
         }

--- a/ncmb-core/src/test/java/com/nifty/cloud/mb/core/NCMBInstallationTest.java
+++ b/ncmb-core/src/test/java/com/nifty/cloud/mb/core/NCMBInstallationTest.java
@@ -554,7 +554,7 @@ public class NCMBInstallationTest {
         out.close();
 
         // check
-        Assert.assertNull(NCMBInstallation.getCurrentInstallation().getObjectId());
+        Assert.assertNotNull(NCMBInstallation.getCurrentInstallation());
     }
 
     //endregion

--- a/ncmb-core/src/test/java/com/nifty/cloud/mb/core/NCMBInstallationTest.java
+++ b/ncmb-core/src/test/java/com/nifty/cloud/mb/core/NCMBInstallationTest.java
@@ -21,6 +21,8 @@ import org.robolectric.shadows.ShadowLog;
 import org.robolectric.shadows.ShadowLooper;
 import org.robolectric.shadows.gms.ShadowGooglePlayServicesUtil;
 
+import java.io.File;
+import java.io.FileOutputStream;
 import java.text.DateFormat;
 
 import static org.mockito.Matchers.any;
@@ -523,6 +525,38 @@ public class NCMBInstallationTest {
         Assert.assertNotNull(error);
         Assert.assertEquals("java.lang.IllegalArgumentException: objectId is must not be null.", error.getMessage());
     }
+
+    /**
+     * - 内容：currentInstallationのローカルファイルが空の状態でgetCurrentInstallationを実行した場合にエラーが発生しないこと
+     * - 結果：エラーが発生しないこと
+     */
+    @Test
+    public void currentInstallation_null() throws Exception {
+        // post
+        NCMBException error = null;
+        NCMBInstallation installation = new NCMBInstallation();
+        installation.setDeviceToken("xxxxxxxxxxxxxxxxxxx");
+        try {
+            installation.save();
+        } catch (NCMBException e) {
+            error = e;
+        }
+        Assert.assertNull(error);
+        Assert.assertEquals("7FrmPTBKSNtVjajm", NCMBInstallation.getCurrentInstallation().getObjectId());
+
+        // clear currentInstallation
+        NCMBInstallation.currentInstallation = null;
+
+        // set 0byte currentInstallation file
+        File file = NCMBLocalFile.create(NCMBInstallation.INSTALLATION_FILENAME);
+        FileOutputStream out = new FileOutputStream(file);
+        out.write("".getBytes("UTF-8"));
+        out.close();
+
+        // check
+        Assert.assertNull(NCMBInstallation.getCurrentInstallation().getObjectId());
+    }
+
     //endregion
 
 

--- a/ncmb-core/src/test/java/com/nifty/cloud/mb/core/NCMBLocalFileTest.java
+++ b/ncmb-core/src/test/java/com/nifty/cloud/mb/core/NCMBLocalFileTest.java
@@ -1,0 +1,97 @@
+package com.nifty.cloud.mb.core;
+
+import com.squareup.okhttp.mockwebserver.MockWebServer;
+
+import junit.framework.Assert;
+
+import org.json.JSONObject;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+import org.junit.runner.RunWith;
+import org.robolectric.Robolectric;
+import org.robolectric.RuntimeEnvironment;
+import org.robolectric.annotation.Config;
+
+import java.io.File;
+
+/**
+ * NCMBLocalFileTest class
+ */
+@RunWith(CustomRobolectricTestRunner.class)
+@Config(constants = BuildConfig.class, sdk = 21, manifest = Config.NONE, shadows = {ShadowNCMBUser.class})
+public class NCMBLocalFileTest {
+
+    private boolean callbackFlag;
+
+    @Before
+    public void setup() throws Exception {
+
+        MockWebServer mServer = new MockWebServer();
+        mServer.setDispatcher(NCMBDispatcher.dispatcher);
+        mServer.start();
+
+        NCMB.initialize(RuntimeEnvironment.application.getApplicationContext(),
+                "appKey",
+                "cliKey",
+                mServer.getUrl("/").toString(),
+                null);
+        NCMBUser.currentUser = null;
+        Robolectric.getBackgroundThreadScheduler().pause();
+        Robolectric.getForegroundThreadScheduler().pause();
+
+        callbackFlag = false;
+    }
+
+    @After
+    public void teardown() {
+
+    }
+
+    /**
+     * TestCase
+     */
+    @Rule
+    public ExpectedException thrown = ExpectedException.none();
+
+    @Test
+    public void create() throws Exception {
+        String fileName = "Test";
+        File file = NCMBLocalFile.create(fileName);
+        Assert.assertEquals(fileName, file.getName());
+    }
+
+    @Test
+    public void writeFile() throws Exception {
+        String jsonString = "{\"key\":\"value\"}";
+
+        File file = NCMBLocalFile.create("Test");
+        Assert.assertFalse(file.exists());
+        NCMBLocalFile.writeFile(file, new JSONObject(jsonString));
+        Assert.assertTrue(file.exists());
+    }
+
+    @Test
+    public void readFile() throws Exception {
+        String jsonString = "{\"key\":\"value\"}";
+        File file = NCMBLocalFile.create("Test");
+        NCMBLocalFile.writeFile(file, new JSONObject(jsonString));
+
+        Assert.assertEquals(jsonString, NCMBLocalFile.readFile(file).toString());
+    }
+
+    @Test
+    public void deleteFile() throws Exception {
+        String jsonString = "{\"key\":\"value\"}";
+        File file = NCMBLocalFile.create("Test");
+
+        NCMBLocalFile.writeFile(file, new JSONObject(jsonString));
+        Assert.assertTrue(file.exists());
+
+        NCMBLocalFile.deleteFile(file);
+        Assert.assertFalse(file.exists());
+    }
+
+}

--- a/ncmb-core/src/test/java/com/nifty/cloud/mb/core/NCMBLocalFileTest.java
+++ b/ncmb-core/src/test/java/com/nifty/cloud/mb/core/NCMBLocalFileTest.java
@@ -16,6 +16,7 @@ import org.robolectric.RuntimeEnvironment;
 import org.robolectric.annotation.Config;
 
 import java.io.File;
+import java.io.FileOutputStream;
 
 /**
  * NCMBLocalFileTest class
@@ -83,6 +84,17 @@ public class NCMBLocalFileTest {
     }
 
     @Test
+    public void readFile_0byte() throws Exception {
+        // set 0byte local file
+        File file = NCMBLocalFile.create("Test");
+        FileOutputStream out = new FileOutputStream(file);
+        out.write("".getBytes("UTF-8"));
+        out.close();
+
+        Assert.assertEquals("{}", NCMBLocalFile.readFile(file).toString());
+    }
+
+    @Test
     public void deleteFile() throws Exception {
         String jsonString = "{\"key\":\"value\"}";
         File file = NCMBLocalFile.create("Test");
@@ -93,5 +105,6 @@ public class NCMBLocalFileTest {
         NCMBLocalFile.deleteFile(file);
         Assert.assertFalse(file.exists());
     }
+
 
 }

